src/main/java/com/bookstore/service/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,129 @@
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final BookRepository bookRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final BookMapper bookMapper;
+
+    @Override
+    public void createShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCart.setId(user.getId());
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto save(UpdateCartItemDto itemDto, User user) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUser(user)
+                .orElseGet(() -> {
+                    ShoppingCart newShoppingCart = new ShoppingCart();
+                    newShoppingCart.setUser(user);
+                    return shoppingCartRepository.save(newShoppingCart);
+                });
+
+        Book book = bookRepository.findById(itemDto.getBookId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Book with ID " + itemDto.getBookId() + " not found")
+                );
+
+        CartItem existingCartItem = shoppingCart.getCartItems().stream()
+                .filter(cartItem -> cartItem.getBook().equals(book))
+                .findFirst()
+                .orElse(null);
+
+        if (existingCartItem != null) {
+            existingCartItem.setQuantity(existingCartItem.getQuantity() + itemDto.getQuantity());
+            cartItemRepository.save(existingCartItem);
+        } else {
+            CartItem newCartItem = new CartItem();
+            newCartItem.setBook(book);
+            newCartItem.setQuantity(itemDto.getQuantity());
+            newCartItem.setShoppingCart(shoppingCart);
+            shoppingCart.getCartItems().add(newCartItem);
+            cartItemRepository.save(newCartItem);
+        }
+
+        shoppingCartRepository.save(shoppingCart);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto getByUserId(Long id) {
+        ShoppingCart shoppingCart = shoppingCartRepository.getByUserId(id)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Shopping cart for user with ID " + id + " not found."
+                ));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public void deleteCartItem(Long cartItemId) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
+                .orElseThrow(() -> new EntityNotFoundException("CartItem with id "
+                        + cartItemId + " not found"));
+        cartItemRepository.delete(cartItem);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addBookToCart(CartItemRequestDto request, User user) {
```

Comment:
What is the difference between addBookToCart and save?

Reply:
same question)